src/compiler/crystal/semantic/restrictions.cr

```diff
@@ -304,6 +304,44 @@ module Crystal
     end
   end
 
+  class GenericClassType
+    def restriction_of?(other : GenericClassInstanceType, owner)
+      # ```
+      # def foo(param : Array)
+      # end
+      #
+      # def foo(param : Array(Int32))
+      # end
+      # ```
+      #
+      # Here, self is `Array`, other is `Array(Int32)`
+
+      return unless self == other.generic_type
+
+      # `SomeGeneric` is never a restriction of `SomeGeneric(X)`
+      false
+    end
+  end
+
+  class GenericClassInstanceType
+    def restriction_of?(other : GenericClassType, owner)
+      # ```
+      # def foo(param : Array(Int32))
+      # end
+      #
+      # def foo(param : Array)
+      # end
+      # ```
+      #
+      # Here, self is `Array(Int32)`, other is `Array`
+
+      return unless self.generic_type == other
```

Comment:
```suggestion
      return false unless self.generic_type == other
```

Reply:
Hum no, that's wrong.. it'd mean that `Foo` is a restriction of `Foo(Int32)` (when the generic type `Foo` is equal)

Reply:
IIUC it's a simplification of your conditions within this method.

Reply:
Oh right, I was thinking about the other one ><

I'll change that!